File: droid/app/src/main/java/com/storyteller/droid/data/KnowledgeSource.kt
package com.storyteller.droid.data

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import java.security.MessageDigest

data class KnowledgeReadCounters(val bytesRead: Long = 0, val chunksOpened: Int = 0, val recordsDecoded: Int = 0)
data class KnowledgeRead(val excerpts: List<KnowledgeEntry>, val counters: KnowledgeReadCounters)

private data class KnowledgeLocator(
    val entryId: String,
    val tokens: Set<String>,
    val revealAfterNodes: Set<String>,
    val path: String,
    val sha256: String,
    val sizeBytes: Long,
)

/** Bounded content-addressed reader for the v2 narrative/knowledge namespace. */
class DirectoryKnowledgeSource(private val root: File) {
    private val gson = Gson()
    private val canonicalRoot = root.canonicalFile
    private val locators: List<KnowledgeLocator> = parseIndex(File(root, "index.json"))

    fun read(
        entryIds: Set<String> = emptySet(),
        queryTokens: Set<String> = emptySet(),
        visitedNodes: Set<String> = emptySet(),
        maxRecords: Int,
        maxExcerptBytes: Long,
    ): KnowledgeRead {
        require(maxRecords >= 0 && maxExcerptBytes >= 0) { "knowledge bounds must be non-negative" }
        if (maxRecords == 0 || maxExcerptBytes == 0L) return KnowledgeRead(emptyList(), KnowledgeReadCounters())
        val excerpts = mutableListOf<KnowledgeEntry>()
        var bytesRead = 0L
        var chunksOpened = 0
        var recordsDecoded = 0
        for (locator in locators) {
            if (excerpts.size == maxRecords) break
            if (entryIds.isNotEmpty() && locator.entryId !in entryIds) continue
            if (queryTokens.isNotEmpty() && locator.tokens.intersect(queryTokens).isEmpty()) continue
            if (!visitedNodes.containsAll(locator.revealAfterNodes)) continue
            if (bytesRead + locator.sizeBytes > maxExcerptBytes) continue
            val file = File(canonicalRoot, locator.path).canonicalFile
            require(file.parentFile?.toPath()?.startsWith(canonicalRoot.toPath()) == true) { "KNOWLEDGE_CHUNK_PATH" }
            require(file.length() == locator.sizeBytes) { "KNOWLEDGE_CHUNK_SIZE" }
            val payload = file.readBytes()
            chunksOpened += 1
            bytesRead += payload.size
            require(sha256(payload) == locator.sha256) { "KNOWLEDGE_CHUNK_HASH" }
            val entry = parseEntry(payload)
            recordsDecoded += 1
            require(entry.entryId == locator.entryId) { "KNOWLEDGE_CHUNK_ID" }
            require(entry.revealAfterNodes.toSet() == locator.revealAfterNodes) { "KNOWLEDGE_CHUNK_REVEAL" }
            excerpts += entry
        }
        return KnowledgeRead(excerpts, KnowledgeReadCounters(bytesRead, chunksOpened, recordsDecoded))
    }

    private fun parseIndex(file: File): List<KnowledgeLocator> {
        val type = object : TypeToken<Map<String, Any>>() {}.type
        val raw: Map<String, Any> = gson.fromJson(file.readText(Charsets.UTF_8), type)
        val entries = raw["entries"] as? List<*> ?: error("KNOWLEDGE_INDEX_FORMAT")
        val parsed = entries.map { value ->
            val item = value as? Map<*, *> ?: error("KNOWLEDGE_INDEX_ENTRY")
            val tokens = strings(item["tokens"])
            val reveal = strings(item["reveal_after_nodes"])
            val path = item["path"] as? String ?: error("KNOWLEDGE_INDEX_ENTRY")
            val hash = item["sha256"] as? String ?: error("KNOWLEDGE_INDEX_ENTRY")
            val size = (item["size_bytes"] as? Number)?.toLong() ?: error("KNOWLEDGE_INDEX_ENTRY")
            require(tokens == tokens.distinct().sorted() && reveal == reveal.distinct().sorted() &&
                !path.startsWith('/') && '\\' !in path && path.split('/').none { it.isEmpty() || it == "." || it == ".." } &&
                hash.matches(Regex("[0-9a-f]{64}")) && size >= 0
            ) { "KNOWLEDGE_INDEX_ENTRY" }
            KnowledgeLocator(
                item["entry_id"] as? String ?: error("KNOWLEDGE_INDEX_ENTRY"),
                tokens.toSet(), reveal.toSet(), path, hash, size,
            )
        }.sortedBy(KnowledgeLocator::entryId)
        require(parsed.map(KnowledgeLocator::entryId).distinct().size == parsed.size) { "KNOWLEDGE_INDEX_DUPLICATE_ID" }
        return parsed
    }

    private fun parseEntry(payload: ByteArray): KnowledgeEntry {
        val type = object : TypeToken<Map<String, Any>>() {}.type
        val item: Map<String, Any> = gson.fromJson(String(payload, Charsets.UTF_8), type)
        return KnowledgeEntry(
            item["entry_id"] as? String ?: error("KNOWLEDGE_CHUNK_FORMAT"),
            item["kind"] as? String ?: error("KNOWLEDGE_CHUNK_FORMAT"),
            item["normalized_text"] as? String ?: error("KNOWLEDGE_CHUNK_FORMAT"),
            strings(item["source_ids"]), strings(item["incoming_refs"]),
            strings(item["outgoing_refs"]), strings(item["reveal_after_nodes"]),
        )
    }

    private fun strings(value: Any?): List<String> =
        (value as? List<*>)?.map { it as? String ?: error("KNOWLEDGE_INDEX_ENTRY") }.orEmpty()

    private fun sha256(payload: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(payload).joinToString("") { "%02x".format(it) }
}

File: ios/StoryTeller/Data/KnowledgeSource.swift
import CryptoKit
import Foundation

struct KnowledgeReadCounters: Equatable {
    let bytesRead: Int
    let chunksOpened: Int
    let recordsDecoded: Int
}

struct KnowledgeRead {
    let excerpts: [KnowledgeEntry]
    let counters: KnowledgeReadCounters
}

private struct KnowledgeLocator {
    let entryId: String
    let tokens: Set<String>
    let revealAfterNodes: Set<String>
    let path: String
    let sha256: String
    let sizeBytes: Int
}

/// Bounded content-addressed reader for the v2 narrative/knowledge namespace.
struct DirectoryKnowledgeSource {
    private let root: URL
    private let locators: [KnowledgeLocator]

    init(root: URL) throws {
        self.root = root.standardizedFileURL
        let data = try Data(contentsOf: root.appendingPathComponent("index.json"))
        guard let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let entries = raw["entries"] as? [[String: Any]] else {
            throw KnowledgeSourceError.invalid("KNOWLEDGE_INDEX_FORMAT")
        }
        var parsed: [KnowledgeLocator] = []
        for item in entries {
            guard let entryId = item["entry_id"] as? String,
                  let tokens = item["tokens"] as? [String],
                  let reveal = item["reveal_after_nodes"] as? [String],
                  let path = item["path"] as? String,
                  let sha256 = item["sha256"] as? String,
                  let sizeBytes = item["size_bytes"] as? Int,
                  tokens == Array(Set(tokens)).sorted(),
                  reveal == Array(Set(reveal)).sorted(),
                  !path.hasPrefix("/"), !path.contains("\\"),
                  path.split(separator: "/", omittingEmptySubsequences: false)
                    .allSatisfy({ !$0.isEmpty && $0 != "." && $0 != ".." }),
                  sha256.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil,
                  sizeBytes >= 0 else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_INDEX_ENTRY")
            }
            parsed.append(KnowledgeLocator(
                entryId: entryId, tokens: Set(tokens), revealAfterNodes: Set(reveal),
                path: path, sha256: sha256, sizeBytes: sizeBytes
            ))
        }
        parsed.sort { $0.entryId < $1.entryId }
        guard Set(parsed.map(\.entryId)).count == parsed.count else {
            throw KnowledgeSourceError.invalid("KNOWLEDGE_INDEX_DUPLICATE_ID")
        }
        locators = parsed
    }

    func read(
        entryIds: Set<String> = [],
        queryTokens: Set<String> = [],
        visitedNodes: Set<String> = [],
        maxRecords: Int,
        maxExcerptBytes: Int
    ) throws -> KnowledgeRead {
        guard maxRecords >= 0, maxExcerptBytes >= 0 else {
            throw KnowledgeSourceError.invalid("knowledge bounds must be non-negative")
        }
        var excerpts: [KnowledgeEntry] = []
        var bytesRead = 0
        var chunksOpened = 0
        var recordsDecoded = 0
        for locator in locators {
            if excerpts.count == maxRecords { break }
            if !entryIds.isEmpty, !entryIds.contains(locator.entryId) { continue }
            if !queryTokens.isEmpty, locator.tokens.isDisjoint(with: queryTokens) { continue }
            if !locator.revealAfterNodes.isSubset(of: visitedNodes) { continue }
            if bytesRead + locator.sizeBytes > maxExcerptBytes { continue }
            let file = root.appendingPathComponent(locator.path).standardizedFileURL
            guard file.path.hasPrefix(root.path + "/") else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_PATH")
            }
            let declaredSize = try file.resourceValues(forKeys: [.fileSizeKey]).fileSize
            guard declaredSize == locator.sizeBytes else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_SIZE")
            }
            let payload = try Data(contentsOf: file, options: .mappedIfSafe)
            chunksOpened += 1
            bytesRead += payload.count
            guard SHA256.hash(data: payload).map({ String(format: "%02x", $0) }).joined() == locator.sha256 else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_HASH")
            }
            guard let item = try JSONSerialization.jsonObject(with: payload) as? [String: Any],
                  let entry = Self.entry(item) else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_FORMAT")
            }
            recordsDecoded += 1
            guard entry.entryId == locator.entryId else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_ID")
            }
            guard Set(entry.revealAfterNodes) == locator.revealAfterNodes else {
                throw KnowledgeSourceError.invalid("KNOWLEDGE_CHUNK_REVEAL")
            }
            excerpts.append(entry)
        }
        return KnowledgeRead(
            excerpts: excerpts,
            counters: KnowledgeReadCounters(
                bytesRead: bytesRead, chunksOpened: chunksOpened, recordsDecoded: recordsDecoded
            )
        )
    }

    private static func entry(_ item: [String: Any]) -> KnowledgeEntry? {
        guard let entryId = item["entry_id"] as? String,
              let kind = item["kind"] as? String,
              let text = item["normalized_text"] as? String,
              let sources = item["source_ids"] as? [String],
              let incoming = item["incoming_refs"] as? [String],
              let outgoing = item["outgoing_refs"] as? [String],
              let reveal = item["reveal_after_nodes"] as? [String] else { return nil }
        return KnowledgeEntry(
            entryId: entryId, kind: kind, normalizedText: text, sourceIds: sources,
            incomingRefs: incoming, outgoingRefs: outgoing, revealAfterNodes: reveal
        )
    }
}

private enum KnowledgeSourceError: Error {
    case invalid(String)
}
